### PR TITLE
[libc++] Fix UB in filesystem::__copy for non-existent destination.

### PR DESCRIPTION
**Pull request by ivafanas**
lstat/stat/fstat has no guarantee whether `struct stat` buffer is changed or not on failure. `filesystem::__copy` function assumes, that `struct stat` buffer is not updated on lstat/stat/fstat failure, which is not true.

It appears that for non-existing destination `detail::posix_lstat(to, t_st, &m_ec1)` returns failure indicator and overwrites `struct stat` buffer with garbage values, which accidentally equal to the `f_st` from stack internals from the previous `detail::posix_lstat(from, f_st, &m_ec1)` call.

`file_type::not_found` is a known status, so check against `if (not status_known(t))` is passed and execution continues further. Then `__copy` function returns `errc::function_not_supported`, because stats are accidentally equivalent, which is a mistake.

Before checking for `detail::stat_equivalent`, need to make sure, that lstat/stat/fstat functions returned success.

As soon as `f_st` and `t_st` access without checking for the lstat/stat/fstat success indicator is an UB, it is not needed to zero-initialize them.

### CODE REVIEW

**Comment by ivafanas** [comment]
@EricWF, could you please review the request.

I do not have commit access. If there are no objections, could you please merge the request into repo.
Thanks.

**Comment by ivafanas** [comment]
Could anybody please review the PR.

**Comment by ivafanas** [comment]
Could anybody please review the PR?
Any comments?

**Comment by 4-rodrigo-salazar** [comment]
I've reviewed this change and it looks good to me. I'm not an approver in this repo though :)

Related, I have gone over the rest of file to find other potentially incorrect error_code paths in https://github.com/llvm/llvm-project/pull/88341 and found a couple more issues (less serious than the bug fix in this diff).

**Comment by ivafanas** [comment]
Ping.

Please, note that the issue is not the theoretical research.
We've actually triggered this bug on some linux distribution.

**Comment by EricWF** [comment]
Are there tests for this? If so, I wonder why MSAN didn't catch it.

Could you please confirm or add a test?

**Comment by ivafanas** [comment]
It is existing test failure:

```
std/input.output/filesystems/fs.op.funcs/fs.op.copy/copy.pass.cpp
TEST_CASE(from_is_regular_file)
```

Issue depends on `fstat/lstat/stat` posix functions implementation.

At first, I thought, that problem is in linux itself, but documentation states that `struct stat` state is undefined if `fstat/lstat/stat` returned "failure" indicator. So, it happens that bug is in `fs::__copy`. It must not assume that `struct stat` state is saved after `fstat/lstat/stat` call.

**Comment by ivafanas** [comment]
Ping.

Failure depends on posix functions implementation, that's why it isn't visible in regular CI.
We've caught the issue on the already existing tests on specific Linux version.

**Comment by ivafanas** [comment]
Hi.
Any comments?

We can reject the ticket if it is unclear.

**Comment by mordante** [comment]
> Hi. Any comments?
> 
> We can reject the ticket if it is unclear.

I'll ping Eric for you. (Missing messages in the GitHub queue is not hard :-/)

**Comment by EricWF** [comment]
> At first, I thought, that problem is in linux itself, but documentation states that `struct stat` state is undefined if `fstat/lstat/stat` returned "failure" indicator.

Which documentation are you referring to?

**Comment by EricWF** [comment]
LGTM once the questions are answered. 

What OS and version are you running specifically to trigger this?

I'll leave it to @mordante to commit.

**Comment by ivafanas** [comment]
Hi!

I've searched for docs here:
https://linux.die.net/man/2/stat
https://man7.org/linux/man-pages/man2/stat.2.html

It does not state that `buf` parameter is not affected if error is returned. It is quite strange to update `buf` on failure, but, well, documentation allows it.

Problem is found on kernel 5.4.0 for the custom linux provided by our customer. Have no idea whether posix implementation of `stat` function is standard or customized.

**Comment by ivafanas** [comment]
Hi!

Let's close this pull request if nobody is interested.

I do not want to track it anymore. We can keep the fix in our customer fork without sharing it to community.

**Review by ldionne — APPROVED** [review]
I went through the PR explanation and the discussions since the start of the PR, and reviewed the code itself.

This seems like a real problem to me, only one that we can't realistically find in our CI testing environments because of how our OSes implement these low level functions. I'm happy to hear that our existing tests would actually catch this issue on an OS where the stat functions are implemented in that particular (but legal) way.

I think there's no reason to hold off with this patch any longer. Thanks for the report, explanation and the fix @ivafanas .

**Pull request merged**
